tests/unit/data/test_storage.py

```diff
@@ -278,3 +285,204 @@ def test_all_data_cached(self):
 
         # noinspection PyUnresolvedReferences
         ds.data_file.get.assert_called_once()
+
+
+class DataSetStoreTest(unittest.TestCase):
+
+    def setUp(self) -> None:
+        self.mock_fs = Mock(spec=S3FileSystem)
+        self.mock_fs.anon = False
+        self.mock_storage_bucket_name = "test_bucket"
+
+        test_metadata = Metadata("test dataset", box(-2, -2, 2, 2))
+
+        self.mock_metadata_store = Mock(spec=S3FSMetadataStore)
+        self.mock_metadata_store.get.return_value = test_metadata
+
+        self.dataset_store = S3FSDataSetStore(self.mock_fs, self.mock_metadata_store, self.mock_storage_bucket_name)
+
+        # This is a little bit circular, but makes it easier to set up tests for download operations
+        # later, since the paths will match
+        data_dir = Path(self.dataset_store._cache_dir) / Path(test_metadata.id)
+        data_files = [data_dir / p for p in [Path("test-file.csv")]]
+        self.test_dataset = DataSet(data_files, test_metadata)
+
+    def test_available_datasets(self):
+        """
+        WHEN available_datasets is called
+        THEN a list of datasets that exist in the storage bucket is returned
+        """
+        expected_datasets = ["dataset1", "dataset2", "dataset3"]
+        self.mock_fs.ls.return_value = [f"{self.mock_storage_bucket_name}/{ds}" for ds in expected_datasets]
+
+        actual_datasets = self.dataset_store.available_datasets()
+        self.assertEqual(expected_datasets, actual_datasets)
+
+    def test_get(self):
+        """
+        GIVEN a valid ID for a dataset that exists on the object store
+        WHEN get is called
+        THEN that dataset is returned
+        """
+
+        def mock_get(_s3_key: str, local_dir: str, **_kwargs):
+            local_dir = Path(local_dir)
+            local_dir.mkdir(parents=True, exist_ok=True)
+            for datafile in self.test_dataset.files:
+                (local_dir / Path(datafile)).touch()
+
+        self.mock_fs.get.side_effect = mock_get
+
+        self.mock_metadata_store.get.return_value = self.test_dataset.metadata
+
+        actual = self.dataset_store.get(self.test_dataset.id)
+
+        self.assertEqual(self.test_dataset, actual)
+        self.mock_metadata_store.get.assert_called_with(self.test_dataset.id)
+        self.mock_fs.get.assert_called_with(
+            f"{self.mock_storage_bucket_name}/{self.test_dataset.id}/",
+            f"{self.dataset_store._cache_dir}/{self.test_dataset.id}",
+            recursive=True
+        )
```

Comment:
However, I see you have multiple asserts in here.  The debate continues...

Reply:
I aim to test a single "thing" per test case, but sometimes validating that "thing" involves multiple asserts.

Here, for example, the "thing" is getting a dataset involves making the expected calls to the metadata store and object store.

Perhaps multiple vs single asserts is a proxy discussion for what is the appropriate scope for a test case? (i.e. what constitutes a "thing")

Reply:
I like the approach of testing one "thing" per test, including as many asserts as is necessary to test that one thing.  But yes, we shall leave this debate for another day.

Reply:
I'm with @crbunney here - in my test I want to confirm a single "thing" (behaviour, perhaps?) but sometimes that behaviour is complex enough to require multiple assertions. You could argue then that the unit you're testing is too large with too complex behaviour, but to follow that path leads to near infinite tiny functions that are impossible to keep track of, which is more distasteful than multiple assertions per test IMO. 

The main reason for pushing for one assertion per test is that a failing assertion early in a test can mask another, potentially more severe failure later on in the test code...